### PR TITLE
Entity Tenant

### PR DESCRIPTION
**Pull request by cervols**
Create model and controller for the Tenant entity. Add validation rules and rspec tests
tenant_edit
tenant_new
tenant_show
tenants_index

### CODE REVIEW

File: app/controllers/tenants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  def new
+    @tenant = Tenant.new
+  end
+
+  def create
+    @tenant = Tenant.new(tenant_params)
+    if @tenant.save
+      redirect_to tenants_path
+    else
+      render 'new'
```
Comment (cervols) [inline app/controllers/tenants_controller.rb]: prefer symbol

File: app/controllers/tenants_controller.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  def edit
+    @tenant = Tenant.find(params[:id])
+  end
+
+  def update
+    @tenant = Tenant.find(params[:id])
+    if @tenant.update_attributes(tenant_params)
+      flash[:success] = "Information is updated"
+      redirect_to @tenant
```
Comment (cervols) [inline app/controllers/tenants_controller.rb]: prefer _url helpers

File: app/controllers/tenants_controller.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    Tenant.find(params[:id]).destroy
+    flash[:success] = "Tenant is deleted"
```
Comment (cervols) [inline app/controllers/tenants_controller.rb]: redirect

File: app/views/tenants/edit.html.slim [hunk 4]
```diff
@@ -0,0 +1,13 @@
+- provide(:button_text, 'Edit Tenant')
+- provide(:path, tenant_path)
```
Comment (cervols) [inline app/views/tenants/edit.html.slim]: you do not need it

File: app/views/tenants/edit.html.slim [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+h1 Edit Tenant
+.row
+  .col-md-6.col-md-offset-3
+    = form_for(@tenant, url: yield(:path)) do |f|
```
Comment (cervols) [inline app/views/tenants/edit.html.slim]: use simple_form gem for forms.

File: app/views/tenants/new.html.slim [hunk 6]
```diff
@@ -0,0 +1,13 @@
+- provide(:button_text, 'Create new Tenant')
+- provide(:path, add_tenant_path)
```
Comment (cervols) [inline app/views/tenants/new.html.slim]: you don't need it.

File: config/routes.rb [hunk 7]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get  '/add_tenant',  to: 'tenants#new'
```
Comment (cervols) [inline config/routes.rb]: read about resources.

Comment (cervols) [inline config/routes.rb]: also read about REST

File: db/migrate/20171210195325_create_tenants.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class CreateTenants < ActiveRecord::Migration[5.1]
+  def change
+    create_table :tenants do |t|
+      t.string :name
+      t.string :phone
+      t.string :website
+      t.string :logo
```
Comment (cervols) [inline db/migrate/20171210195325_create_tenants.rb]: for files use carrier_wave

File: db/seeds.rb [hunk 9]
```diff
@@ -1,7 +1,7 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+Tenant.create!(name:  'First Tenant',
```
Comment (cervols) [inline db/seeds.rb]: You could use factory.

File: app/views/tenants/edit.html.slim (hunk 5 above)
Comment (egorodessa) [reply app/views/tenants/edit.html.slim]: как эта штука работает? мне кажеться что ты сама с собой разговариваешь, это так?

Comment (cervols) [reply app/views/tenants/edit.html.slim]: I like talking to an intelligent person ;)

File: app/models/tenant.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class Tenant < ApplicationRecord
+  VALID_WEBSITE_REGEX = /\A(www.)?[^_\W][-a-zA-Z0-9_]+\.+[-a-zA-Z0-9]+\z/i
+  VALID_PHONE_REGEX = /\A[+]?[\d\-.() ]+\z/
+
+  has_many :users
+  has_many :companies, through: :subscriptions
```
Comment (igorkasyanchuk) [inline app/models/tenant.rb]: through: :subscriptions ?? what is subscriptions?

File: app/views/account/tenants/show.html.slim [hunk 11]
```diff
@@ -0,0 +1,15 @@
+h1= @tenant.name
+.btn-group
+  = link_to  "Edit", edit_account_tenant_path, class: 'btn btn-primary'
+  = link_to "Delete", account_tenant_path, method: :delete,
+                      data: { confirm: 'Are you sure you want to delete this tenant?' },
+                      class: 'btn btn-primary'
+.tenant
+  p
```
Comment (igorkasyanchuk) [inline app/views/account/tenants/show.html.slim]: this is something what must be improved in future. such HTML to show details is bad.

File: app/views/layouts/application.html.slim [hunk 12]
```diff
@@ -10,5 +10,6 @@ html
     p.notice
       = notice
     p.alert
-      = alert
-    = yield
+      - flash.each do |message_type, message|
```
Comment (igorkasyanchuk) [inline app/views/layouts/application.html.slim]: I see issue here with code ...
we have above.
p.notice
p.alert
and now we have flash.each

just leave flash.each in div. Remove "p"

File: db/migrate/20171210195325_create_tenants.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+class CreateTenants < ActiveRecord::Migration[5.1]
+  def change
+    create_table :tenants do |t|
+      t.string :name
```
Comment (igorkasyanchuk) [inline db/migrate/20171210195325_create_tenants.rb]: tenant belongs to user. add this relation
t.integer :owner_id, index: true

class Tenant
  has_one :owner, class_name: User

File: spec/controllers/tenant_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    sign_in @user
+  end
+  let!(:tenant) { FactoryBot.create(:tenant) }
+
+  it 'Get#index' do
+    get :index
+    expect(response).to have_http_status(200)
+  end
+
+  it 'Get#show' do
```
Comment (igorkasyanchuk) [inline spec/controllers/tenant_controller_spec.rb]: change to "GET #show", similar to other specs

File: app/views/account/tenants/edit.html.slim [hunk 15]
```diff
@@ -0,0 +1,9 @@
+h1.col-md-6.col-md-offset-3 Edit Tenant
+.row
+  .col-md-3.col-md-offset-3
+    = simple_form_for @tenant, url: account_tenant_path,
+                                defaults: { input_html: { class: 'form-control' } } do |f|
+      = f.input :name
+      = f.input :website, hint: 'Without "http://" or "https://"'
```
Comment (igorkasyanchuk) [inline app/views/account/tenants/edit.html.slim]: this form MUST be extracted to partial. Do not repeat yourself (DRY).
I see same code in NEW view

File: app/views/account/tenants/index.html.slim [hunk 16]
```diff
@@ -0,0 +1,8 @@
+h1 All tenants
+= link_to "Add new tenant", new_account_tenant_path, class: 'btn btn-primary'
+
+ul.tenant
```
Comment (igorkasyanchuk) [inline app/views/account/tenants/index.html.slim]: remove this class "ul.tenant". leave just "ul"

File: app/views/account/tenants/new.html.slim [hunk 17]
```diff
@@ -0,0 +1,10 @@
+h1.col-md-6.col-md-offset-3 Add new Tenant
+.row
+  .col-md-3.col-md-offset-3
+    = simple_form_for @tenant, url: account_tenants_path,
+                                defaults: { input_html: { class: 'form-control' } } do |f|
+      = f.input :name, placeholder: 'Enter Tenant name'
+      = f.input :website, placeholder: 'domain.com',
+                         hint: 'Without "http://" or "https://"'
+      = f.input :phone, placeholder: '+00 00 00 00'
```
Comment (igorkasyanchuk) [inline app/views/account/tenants/new.html.slim]: remove placeholder or put better (with US number format)

File: app/views/account/tenants/new.html.slim [hunk 18]
```diff
@@ -0,0 +1,10 @@
+h1.col-md-6.col-md-offset-3 Add new Tenant
+.row
+  .col-md-3.col-md-offset-3
+    = simple_form_for @tenant, url: account_tenants_path,
+                                defaults: { input_html: { class: 'form-control' } } do |f|
+      = f.input :name, placeholder: 'Enter Tenant name'
+      = f.input :website, placeholder: 'domain.com',
+                         hint: 'Without "http://" or "https://"'
```
Comment (igorkasyanchuk) [inline app/views/account/tenants/new.html.slim]: I suggest removing this hint

File: app/controllers/account/tenants_controller.rb [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  end
+
+  def edit
+    @tenant = current_tenant
+  end
+
+  def update
+    @tenant = current_tenant
+    if @tenant.update_attributes(tenant_params)
+      flash[:success] = "Information is updated"
```
Comment (romaslmd) [inline app/controllers/account/tenants_controller.rb]: fyi,
you may add custom flash type and use it within `#redirect_to` as option, e.g.:
```
class TenantsController < ApplicationController
  add_flash_types :success

  def update
    if @tenant.update(tenant_params)
      redirect_to account_tenant_path(@tenant), success: 'Information updated successfully'
    else
      render :edit
    end
  end
end
```

File: app/models/tenant.rb (hunk 10 above)
Comment (cervols) [reply app/models/tenant.rb]: changed to through: :users

File: app/views/account/tenants/edit.html.slim (hunk 15 above)
Comment (cervols) [reply app/views/account/tenants/edit.html.slim]: done

File: app/views/account/tenants/index.html.slim (hunk 16 above)
Comment (cervols) [reply app/views/account/tenants/index.html.slim]: done

File: app/views/account/tenants/new.html.slim (hunk 18 above)
Comment (cervols) [reply app/views/account/tenants/new.html.slim]: done

File: app/views/account/tenants/new.html.slim (hunk 17 above)
Comment (cervols) [reply app/views/account/tenants/new.html.slim]: placeholder was removed

File: app/views/account/tenants/show.html.slim (hunk 11 above)
Comment (cervols) [reply app/views/account/tenants/show.html.slim]: I agree. I will improve this code in the future

File: app/views/layouts/application.html.slim (hunk 12 above)
Comment (cervols) [reply app/views/layouts/application.html.slim]: It is true. Now we haven't this issue

File: db/migrate/20171210195325_create_tenants.rb (hunk 13 above)
Comment (cervols) [reply db/migrate/20171210195325_create_tenants.rb]: done

File: spec/controllers/tenant_controller_spec.rb (hunk 14 above)
Comment (cervols) [reply spec/controllers/tenant_controller_spec.rb]: done

File: app/controllers/account/tenants_controller.rb (hunk 19 above)
Comment (cervols) [reply app/controllers/account/tenants_controller.rb]: Thank you for your suggest. I believe I'll can do it after merging my branch with master

**Pull request merged**